Eliminate calls to bundleForClass: at auxiliary process launch time

bnham opened the pull request:
#### 17786394fdc1af61ab7521d9cd8322b848a7ce1a
<pre>
Eliminate calls to bundleForClass: at auxiliary process launch time
<a href="https://bugs.webkit.org/show_bug.cgi?id=269874">https://bugs.webkit.org/show_bug.cgi?id=269874</a>
<a href="https://rdar.apple.com/120925002">rdar://120925002</a>

Reviewed by Per Arne Vollan.

Some perf folks want us to eliminate calls to `+[NSBundle bundleForClass:]` that we do at auxiliary
process launch time. To do this, pass the bundle version to the compiler via a preprocessor macro
rather than looking it up dynamically via NSBundle at runtime.

* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::isSystemWebKit):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::checkFrameworkVersion):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::isSystemWebKit): Deleted.
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::shouldAllowNonValidInjectedCode const):

Canonical link: <a href="https://commits.webkit.org/275405@main">https://commits.webkit.org/275405@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e72f0598ecbfdfdec8f466dfef71e83e5663ae1e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44130 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37841 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18093 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34505 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15175 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15383 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45713 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37936 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37297 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41048 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16563 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13601 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39485 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18182 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9359 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18239 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17826 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan reviewed (approved):
R=me.

emw-apple commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -346,11 +346,10 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
+#if !defined(WEBKIT_IS_PRODUCTION_BUILD)
+    xpc_dictionary_set_fd(bootstrapMessage.get(), "stdout", STDOUT_FILENO);
+    xpc_dictionary_set_fd(bootstrapMessage.get(), "stderr", STDERR_FILENO);
+#endif
```
This is a behavior change when, for instance, an Apple engineer downloads a production root and converts it into a relocatable bundle. Also, STP and downlevel roots are "production" but not installed to /System. So I don't think you can assume that production implies "system WebKit".

bnham replied:
Yeah I am aware of that. I think the question is more whether we want a production root to be telling the auxiliary process to forward stdout/stderr to UIProcess's stdout/stderr or not.

I could pretty easily just keep this logic the same without the WEBKIT_IS_PRODUCTION_BUILD thing, let me take a look into it.

emw-apple commented on Source/WebKit/Configurations/BaseTarget.xcconfig:
```diff
@@ -31,8 +31,12 @@ SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRA
 WK_FORMATREADERBUNDLE_CODESIGNING_DEFINES = ENABLE_LOWER_FORMATREADERBUNDLE_CODESIGNING_REQUIREMENTS;
 WK_FORMATREADERBUNDLE_CODESIGNING_DEFINES[config=Production] = ;
 
+WK_BUNDLE_VERSION_DEFINES = WEBKIT_BUNDLE_VERSION='"$(BUNDLE_VERSION)"';
```
nit: Does the quoting here accomplish anything? It usually doesn't — nothing in the build system is doing shell expansion.

bnham replied:
It's definitely required due to the way `-D` works and the way `GCC_PREPROCESSOR_DEFINITIONS` works with it. You need to convince Xcode to pass the extra double quotes to `-D` for this to actually be defined as a string literal. And I'm pretty sure the only way to convince Xcode to pass those extra double quotes is by surrounding them with single quotes.

Consider:

```
$ cat foo.c
#include <stdio.h>

int main() {
  printf("%s\n", FOO);
}
```

Without double quotes:

```
$ clang '-DFOO=1.2.34+' -o foo foo.c
foo.c:4:18: error: invalid suffix '.34' on floating constant
  printf("%s\n", FOO);
                 ^
<command line>:1:16: note: expanded from macro 'FOO'
#define FOO 1.2.34+
               ^
foo.c:4:21: error: expected expression
  printf("%s\n", FOO);
                    ^
2 errors generated.
```

With double quotes:

```
$ clang '-DFOO="1.2.34+"' -o foo foo.c
<no error>
```

emw-apple replied:
I see, it's the _preprocessor_ that needs to see those quotes (so that it parses as a string literal). Thanks for clarifying!

emw-apple reviewed (changes requested):
I have some questions about WEBKIT_IS_PRODUCTION_BUILD, but otherwise this approach seems fine to me.

bnham committed:
Eliminate calls to bundleForClass: at auxiliary process launch time
https://bugs.webkit.org/show_bug.cgi?id=269874
rdar://120925002

Reviewed by Per Arne Vollan.

Some perf folks want us to eliminate calls to `+[NSBundle bundleForClass:]` that we do at auxiliary
process launch time. To do this, pass the bundle version to the compiler via a preprocessor macro
rather than looking it up dynamically via NSBundle at runtime.

* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/AuxiliaryProcessCocoa.mm:
(WebKit::AuxiliaryProcess::isSystemWebKit):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::checkFrameworkVersion):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::isSystemWebKit): Deleted.
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::finishLaunchingProcess):
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::shouldAllowNonValidInjectedCode const):

Canonical link: https://commits.webkit.org/275405@main
